### PR TITLE
[Spec] More precise checks for Reified Type Parameters

### PR DESCRIPTION
**Pull request by abreslav**
Some issues to be covered by this spec:

https://youtrack.jetbrains.com/issue/KT-6484
https://youtrack.jetbrains.com/issue/KT-3972
https://youtrack.jetbrains.com/issue/KT-7337
https://youtrack.jetbrains.com/issue/KT-7352

### CODE REVIEW

File: spec-docs/reified-type-parameters.md [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+inline fun foo<reified T>() {}
+```
+
+## Semantics, Checks and Restrictions
+
+**Definition** A well-formed type is called *runtime-available* if
+- it has the form `C`, where `C` is a classifier (object, class or trait) that has either no type parameters, or all its type parameters are `reified`, with the exception for class `Nothing`,
+- it has the form `G<A1, ..., An>`, where `G` is a classifier with `n` type parameters, and for every type parameter `Ti` at least one of the following conditions hold:
+    - `Ti` is a `reified` type parameter and the corresponding type argument `Ai` is a runtime-available type,
```
Comment (VladimirReshetnikov) [inline spec-docs/reified-type-parameters.md]: Can a type parameter of a classifier (other than `Array<>`) be `reified`?

**Comment by udalov** [comment]
Isn't this in master already?

**Comment by ilya-g** [comment]
Would it be possible having a value of `Iterable<T>` to check whether it is an instance of `List<T>`?

**Comment by abreslav** [comment]
Yes, when it is exactly the same `T` on both sides. This is already supported, btw, and you don't even have to specify `T` explicitly

**Comment by abreslav** [comment]
Force-pushed the actual diff with `master`

**Comment by ilya-g** [comment]
Prohibiting unchecked cast to non-runtime-available types would render some of our APIs impossible

**Comment by abreslav** [comment]
@ilya-g You are right, it should be a warning, not an error

**Pull request closed without merging**